File: plugin/src/main/java/com/oldstats/tracking/PetTracker.java
package com.oldstats.tracking;

import com.oldstats.api.OldStatsApiClient;
import com.oldstats.api.StatEvent;
import java.util.List;
import net.runelite.api.ChatMessageType;
import net.runelite.api.Client;
import net.runelite.api.NPC;
import net.runelite.api.Player;
import net.runelite.api.coords.WorldPoint;
import net.runelite.api.events.ChatMessage;
import net.runelite.api.events.GameTick;
import net.runelite.api.events.NpcSpawned;

/**
 * There's no dedicated "pet received" event, and the chat message doesn't
 * name the pet. This watches for one of the three known pet-drop messages,
 * then attributes it to whichever NPC spawns next to the player within a
 * few ticks (pets that auto-follow spawn this way). Pets received directly
 * to the inventory instead (follower slot already occupied, or no space)
 * won't have a name resolved and are reported as "Unknown pet".
 */
public class PetTracker {
    private static final int PET_WAIT_TICKS = 3;
    private static final List<String> PET_MESSAGES = List.of(
        "you have a funny feeling like you're being followed",
        "you feel something weird sneaking into your backpack"
    );

    private final OldStatsApiClient apiClient;
    private final Client client;

    private Integer pendingSince;
    private int tick = 0;

    public PetTracker(OldStatsApiClient apiClient, Client client) {
        this.apiClient = apiClient;
        this.client = client;
    }

    public void onChatMessage(ChatMessage event) {
        if (event.getType() != ChatMessageType.GAMEMESSAGE && event.getType() != ChatMessageType.SPAM) return;
        String message = event.getMessage().toLowerCase();
        if (PET_MESSAGES.stream().anyMatch(message::contains)) {
            pendingSince = tick;
        }
    }

    public void onNpcSpawned(NpcSpawned event) {
        Integer since = pendingSince;
        if (since == null) return;
        if (tick - since > PET_WAIT_TICKS) {
            pendingSince = null;
            apiClient.enqueue(new StatEvent.PetReceived(null));
            return;
        }
        NPC npc = event.getNpc();
        Player localPlayer = client.getLocalPlayer();
        WorldPoint playerLoc = localPlayer != null ? localPlayer.getWorldLocation() : null;
        if (playerLoc == null) return;
        WorldPoint npcLoc = npc.getWorldLocation();
        if (npcLoc == null) return;
        if (npcLoc.distanceTo(playerLoc) <= 1) {
            pendingSince = null;
            apiClient.enqueue(new StatEvent.PetReceived(npc.getName()));
        }
    }

    public void onGameTick(GameTick event) {
        tick++;
        Integer since = pendingSince;
        if (since == null) return;
        if (tick - since > PET_WAIT_TICKS) {
            pendingSince = null;
            apiClient.enqueue(new StatEvent.PetReceived(null));
        }
    }
}

File: plugin/src/main/kotlin/com/oldstats/tracking/PetTracker.kt
package com.oldstats.tracking

import com.oldstats.api.OldStatsApiClient
import com.oldstats.api.StatEvent
import net.runelite.api.Client
import net.runelite.api.ChatMessageType
import net.runelite.api.events.ChatMessage
import net.runelite.api.events.GameTick
import net.runelite.api.events.NpcSpawned

/**
 * There's no dedicated "pet received" event, and the chat message doesn't
 * name the pet. This watches for one of the three known pet-drop messages,
 * then attributes it to whichever NPC spawns next to the player within a
 * few ticks (pets that auto-follow spawn this way). Pets received directly
 * to the inventory instead (follower slot already occupied, or no space)
 * won't have a name resolved and are reported as "Unknown pet".
 */
class PetTracker(
    private val apiClient: OldStatsApiClient,
    private val client: Client,
) {
    private var pendingSince: Int? = null
    private var tick = 0

    companion object {
        private const val PET_WAIT_TICKS = 3
        private val PET_MESSAGES = listOf(
            "you have a funny feeling like you're being followed",
            "you feel something weird sneaking into your backpack",
        )
    }

    fun onChatMessage(event: ChatMessage) {
        if (event.type != ChatMessageType.GAMEMESSAGE && event.type != ChatMessageType.SPAM) return
        val message = event.message.lowercase()
        if (PET_MESSAGES.any { message.contains(it) }) {
            pendingSince = tick
        }
    }

    fun onNpcSpawned(event: NpcSpawned) {
        val since = pendingSince ?: return
        if (tick - since > PET_WAIT_TICKS) {
            pendingSince = null
            apiClient.enqueue(StatEvent.PetReceived(petName = null))
            return
        }
        val npc = event.npc
        val playerLoc = client.localPlayer?.worldLocation ?: return
        val npcLoc = npc.worldLocation ?: return
        if (npcLoc.distanceTo(playerLoc) <= 1) {
            pendingSince = null
            apiClient.enqueue(StatEvent.PetReceived(petName = npc.name))
        }
    }

    fun onGameTick(@Suppress("UNUSED_PARAMETER") event: GameTick) {
        tick++
        val since = pendingSince ?: return
        if (tick - since > PET_WAIT_TICKS) {
            pendingSince = null
            apiClient.enqueue(StatEvent.PetReceived(petName = null))
        }
    }
}
